rllib/evaluation/collectors/simple_list_collector.py

```diff
@@ -232,131 +234,69 @@ def build(self, view_requirements: ViewRequirementsDict) -> SampleBatch:
                     _to_float_np_array(d) for d in self.buffers[data_col]
                 ]
 
-            # Range of indices on time-axis, e.g. "-50:-1". Together with
-            # the `batch_repeat_value`, this determines the data produced.
-            # Example:
-            #  batch_repeat_value=10, shift_from=-3, shift_to=-1
-            #  buffer=[-3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
-            #  resulting data=[[-3, -2, -1], [7, 8, 9]]
-            #  Range of 3 consecutive items repeats every 10 timesteps.
-            if view_req.shift_from is not None:
-                # Batch repeat value > 1: Only repeat the shift_from/to range
-                # every n timesteps.
-                if view_req.batch_repeat_value > 1:
-                    count = int(
-                        math.ceil(
-                            (len(np_data[data_col][0]) - self.shift_before)
-                            / view_req.batch_repeat_value
-                        )
+            assert (
+                view_req.shift_arr is not None
+            ), "View requirement shift_arr cannot be None."
+            data = []
+
+            # Go throught each time-step in the buffer and construct the view
+            # accordingly.
+            for d in np_data[data_col]:
+                shifted_data = []
+
+                # batch_repeat_value determines how many time steps should we skip
+                # before we repeat indexing the data.
+                # Example: batch_repeat_value=10, shift_arr = [-3, -2, -1],
+                # shift_before = 3
+                # buffer = [-3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
+                # resulting_data = [[-3, -2, -1], [7, 8, 9]]
+                # explanation: For t=0, we output [-3, -2, -1]. We then skip 10 time
+                # steps ahead and get to t=10. For t=10, we output [7, 8, 9]. We skip
+                # 10 more time steps and get to t=20. but since t=20 is out of bound we
+                # stop.
+
+                # count computes the number of time steps that we need to consider.
+                # if batch_repeat_value = 1, this number should be the length of
+                # episode so far, which is len(buffer) - shift_before.
+                count = int(
+                    math.ceil(
```

Comment:
man, I actually understand batch_repeat now, with your comments :)
one quick question, why is this math.ceil()? I would imagine if we don't want to go out of bounds, this should be math.floor()?

Reply:
I just copied this part from the old logic. But here is why I think math.ceil is actually correct: The way it's used later is with range(ceil) which excludes the count value itself. So, in the worst case you'll hit `len(d) - self.shift_before` which is the true value of the episode length and your i would go from `0` to `len(d) - self.shift_before - 1` which correct.

Reply:
oh, that's true. good point!